benchmarks/generate_datasets.cc

```diff
@@ -0,0 +1,114 @@
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+const char *file_prefix = "dataset.";
+const char *file_suffix = ".pb";
+
+#include <fstream>
+#include <iostream>
+#include "benchmarks.pb.h"
+#include "google_message1.h"
+#include "google_message2.h"
+
+using benchmarks::BenchmarkDataset;
+using google::protobuf::Descriptor;
+using google::protobuf::DescriptorPool;
+using google::protobuf::Message;
+using google::protobuf::MessageFactory;
+
+#define ARRAY_TO_STRING(arr) std::string(arr, arr + sizeof(arr))
+
+std::set<std::string> names;
+
+void WriteFileWithPayloads(const std::string& name,
+                           const std::string& message_name,
+                           const std::vector<std::string>& payload) {
+  if (!names.insert(name).second) {
+    std::cerr << "Duplicate test name: " << name << "\n";
+    abort();
+  }
+
+  // First verify that this message name exists in our set of benchmark messages
+  // and that these payloads are valid for the given message.
+  const Descriptor* d =
+      DescriptorPool::generated_pool()->FindMessageTypeByName(message_name);
+
+  if (!d) {
+    std::cerr << "For dataset " << name << ", no such message: "
+              << message_name << "\n";
+    abort();
+  }
+
+  Message* m = MessageFactory::generated_factory()->GetPrototype(d)->New();
+
+  for (size_t i = 0; i < payload.size(); i++) {
+    if (!m->ParseFromString(payload[i])) {
+      std::cerr << "For dataset " << name << ", payload[" << i << "] fails "
+                << "to parse\n";
+      abort();
+    }
+  }
+
+  BenchmarkDataset dataset;
+  dataset.set_name(name);
+  dataset.set_message_name(message_name);
+  for (size_t i = 0; i < payload.size(); i++) {
+    dataset.add_payload()->assign(payload[i]);
+  }
+
+  std::string serialized;
+  dataset.SerializeToString(&serialized);
+
+  std::ofstream writer;
+  std::string fname = file_prefix + name + file_suffix;
+  writer.open(fname);
+  writer << serialized;
+  writer.close();
+
+  std::cerr << "Wrote dataset: " << fname << "\n";
+}
+
+void WriteFile(const std::string& name, const std::string& message_name,
+               const std::string& payload) {
+  std::vector<std::string> payloads;
+  payloads.push_back(payload);
+  WriteFileWithPayloads(name, message_name, payloads);
+}
+
+int main() {
+  WriteFile("google_message1_proto3", "benchmarks.p3.GoogleMessage1",
+            ARRAY_TO_STRING(google_message1_dat));
```

Comment:
The Makefile is running `xxd` to turn the `.dat` files into `.h` files that declare character arrays. The emitted `.h` file looks like this:

```c
unsigned char google_message1_dat[] = {
  0x0a, 0x00, 0x10, 0x08, 0x18, 0xcb, 0x8f, 0x7e, 0x22, 0x06, 0x33, 0x4b,
  // ...
  0x7e, 0x23, 0x6e, 0x7b, 0x98, 0x04, 0x88, 0x91, 0x61, 0xa0, 0x06, 0x1f
};
unsigned int google_message1_dat_len = 228;
```

I embedded the test messages into the binary instead of loading the files at runtime so that the dataset generating program would be more robust (and not sensitive to the directory you run it from, for example). I always hate programs that seem to get confused unless you run them in just the right way.